src/libretro/libretro.c

```diff
@@ -0,0 +1,534 @@
+/*
+libretro.c - RVVM libretro core
+Copyright (C) 2023  宋文武 <iyzsong@envs.net>
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <libgen.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include "libretro.h"
+
+// RVVM headers
+#include "rvvmlib.h"
+#include "utils.h"
+#include "threading.h"
+#include "devices/clint.h"
+#include "devices/plic.h"
+#include "devices/syscon.h"
+#include "devices/rtc-goldfish.h"
+#include "devices/pci-bus.h"
+#include "devices/nvme.h"
+#include "devices/framebuffer.h"
+#include "devices/i2c-oc.h"
+#include "devices/hid_api.h"
+
+
+static void fallback_log(enum retro_log_level level, const char *fmt, ...);
+static retro_log_printf_t log_cb = fallback_log;
+static retro_video_refresh_t video_cb;
+static retro_input_poll_t input_poll_cb;
+static retro_input_state_t input_state_cb;
+static retro_environment_t environ_cb;
+static rvvm_machine_t *machine;
+static thread_ctx_t *vm_thread;
+static fb_ctx_t vm_fb;
+static hid_keyboard_t *vm_keyboard;
+static hid_mouse_t *vm_mouse;
+static struct {
+    int smp;
+    int mem;
```

Comment:
This causes UB and unexpected issues when setting RAM to >=2048M (It overflows on shift later on). Use `size_t` here (Same goes to cores?).

Rule of thumb: never use signed types unless it is mandatory to handle negative values. For anything that counts objects or size the go-to type is `size_t`.

It might be wise to handle the size suffixes in parser

Reply:
Done, thank you.